### PR TITLE
[RLlib] Add "shuffle batch per epoch" option.

### PR DESCRIPTION
**Pull request by sven1977**
Add "shuffle batch per epoch" option.
* For PPO and any other algo using minibatching AND more than 1 epochs AND  per train batch.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/learner.py [hunk 1]
```diff
@@ -1300,22 +1318,21 @@ def _update_from_batch_or_episodes(
         if minibatch_size:
             if self._learner_connector is not None:
                 batch_iter = partial(
-                    MiniBatchCyclicIterator,
-                    uses_new_env_runners=True,
-                    num_total_mini_batches=num_total_mini_batches,
+                    MiniBatchCyclicIterator, _uses_new_env_runners=True
```
Comment (simonsays1980) [inline rllib/core/learner/learner.py:1321]: Do we need this here still? I thought we deprecate the hybrid stack?

File: rllib/utils/minibatch_utils.py [hunk 2]
```diff
@@ -9,75 +9,94 @@
...
 
-        self._uses_new_env_runners = uses_new_env_runners
+        self._uses_new_env_runners = _uses_new_env_runners
 
-        self._mini_batch_count = 0
-        self._num_total_mini_batches = num_total_mini_batches
+        self._minibatch_count = 0
+        self._num_total_minibatches = num_total_minibatches
 
     def __iter__(self):
```
Comment (simonsays1980) [inline rllib/utils/minibatch_utils.py:88]: In the long run we might want to override `DataIterator` from `ray.data` to build batches from `MultiAgentEpisode`s. Less code.

File: rllib/utils/minibatch_utils.py [hunk 3]
```diff
@@ -140,6 +159,11 @@ def get_len(b):
                     n_steps -= len_sample
                     s = 0
                     self._num_covered_epochs[module_id] += 1
```
Comment (simonsays1980) [inline rllib/utils/minibatch_utils.py:161]: This is actually the same logic like `independent` sampling mode in our `MultiAgentEpisodeBuffer`s. For the future we should reduce code again.

File: rllib/utils/tests/test_minibatch_utils.py [hunk 4]
```diff
@@ -14,20 +14,20 @@
...
-    {"mini_batch_size": 128, "num_sgd_iter": 10, "agent_steps": (1000, 2)},
-    {"mini_batch_size": 128, "num_sgd_iter": 3, "agent_steps": (56, 56)},
-    {"mini_batch_size": 128, "num_sgd_iter": 7, "agent_steps": (56, 56)},
-    {"mini_batch_size": 128, "num_sgd_iter": 10, "agent_steps": (56, 56)},
-    {"mini_batch_size": 128, "num_sgd_iter": 10, "agent_steps": (56, 3)},
-    {"mini_batch_size": 128, "num_sgd_iter": 10, "agent_steps": (56, 4)},
-    {"mini_batch_size": 128, "num_sgd_iter": 10, "agent_steps": (56, 55)},
-    {"mini_batch_size": 128, "num_sgd_iter": 10, "agent_steps": (400, 400)},
-    {"mini_batch_size": 128, "num_sgd_iter": 10, "agent_steps": (64, 64)},
+    {"minibatch_size": 256, "num_epochs": 30, "agent_steps": (1652, 1463)},
```
Comment (simonsays1980) [inline rllib/utils/tests/test_minibatch_utils.py:17]: Nice test!

**Review by simonsays1980 — APPROVED** [review]
LGTM. For the future we might need to inspect critically if we can reuse code we have written elsewhere and if we can replace a lot of iterating by `ray.data` dataset iteration. This reduces code further and places the logic where it belongs, i.e. iterating through data to `ray.data`.

File: rllib/utils/minibatch_utils.py (hunk 3 above)
Comment (sven1977) [reply rllib/utils/minibatch_utils.py:161]: Ah, great catch! We'll get rid of this iterator anyways b/c it relies on `MultiAgentBatch`. We should rewrite it to rely only on a plain dict.

File: rllib/core/learner/learner.py (hunk 1 above)
Comment (sven1977) [reply rllib/core/learner/learner.py:1321]: Officially, not yet. PR is still pending ...

File: rllib/algorithms/algorithm_config.py [hunk 5]
```diff
@@ -2103,6 +2113,15 @@ def training(
                 stack, this setting should no longer be used. Instead, use
                 `train_batch_size_per_learner` (in combination with
                 `num_learners`).
+            num_epochs: The number of complete passes over the entire train batch (per
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:2135]: Will add!

File: rllib/algorithms/appo/appo.py [hunk 6]
```diff
@@ -185,7 +188,7 @@ def training(
             target_network_update_freq: The frequency to update the target policy and
                 tune the kl loss coefficients that are used during training. After
                 setting this parameter, the algorithm waits for at least
-                `target_network_update_freq * minibatch_size * num_sgd_iter` number of
+                `target_network_update_freq * minibatch_size * num_epochs` number of
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/appo/appo.py:191]: Oh wait, great catch. I think this comment here is incorrect.
When we update e.g. PPO with a batch of 4000, the num_env_steps_trained_lifetime counter only(!) gets increased by that 4000, and NOT by: num_epochs * 4000. So for APPO here, this is also wrong. Will fix the comment and clarify.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/appo/appo.py:191]: fixed

File: rllib/algorithms/marwil/marwil.py [hunk 7]
```diff
@@ -398,7 +398,7 @@ class (multi-/single-learner setup) and evaluation on
             learner_results = self.learner_group.update_from_batch(
                 batch,
                 minibatch_size=self.config.train_batch_size_per_learner,
-                num_iters=self.config.dataset_num_iters_per_learner,
+                num_epochs=self.config.dataset_num_iters_per_learner,
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/marwil/marwil.py:401]: Great catch. Clarified the arg names and added this protection to `Learner.update_from_iterator` for now:
```
        if "num_epochs" in kwargs:
            raise ValueError(
                "`num_epochs` arg NOT supported by Learner.update_from_iterator! Use "
                "`num_iters` instead."
            )
```
such that it cannot be confused with `num_epochs` passed in by accident.
